Llamafiler HTTP server stability/reliability fixes

vlasky opened the pull request:
1. Remove aggressive "cancel oldest client" logic in Worker::begin() 

Previously, when all worker threads were busy, the code would forcibly cancel the oldest active connection to make room for a new one. This approach:

* Prematurely terminates in-flight requests, leading to broken or truncated responses.
* Forces cleanup of file descriptors mid-stream, causing spurious "Illegal seek"/EBADF errors.
* Circumvents the TCP backlog queuing provided by the OS, instead dropping live clients and degrading user experience.

By deleting this block, we let the kernel's listen backlog naturally queue incoming connections until a worker becomes available. As a result:

* Active requests are no longer killed arbitrarily under load.
* File descriptors aren't closed unexpectedly, eliminating related "static asset pread" failures.
* The server benefits from standard TCP connection handling without manual interference.

This change improves reliability under high concurrency and avoids unintended side effects from thread cancellation.

2. Fix partial write handling in send_binary()

The current send_binary() implementation treats any write() that returns
less than the requested byte count as a failure, immediately setting
close_connection_ = true and returning false.

This is incorrect behavior. Per POSIX, write() may legitimately return
fewer bytes than requested when:
- The socket send buffer is nearly full
- Network congestion causes backpressure
- Large write sizes exceed kernel buffers
- System is under memory pressure

These partial writes are normal, not error conditions. The current code
incorrectly drops active connections during normal operation, especially
under load when partial writes become more common.

This commit replaces the single write() call with a proper write loop
that:
- Accumulates partial writes until all data is sent
- Retries on EAGAIN/EWOULDBLOCK without closing the connection
- Only treats actual errors (not partial writes) as failures
- Maintains the existing error logging behavior

This fix prevents spurious connection drops during large responses or
when the network is congested, significantly improving reliability under
production load.

Fixes: Connection drops during static file serving
Fixes: Increased failure rate under high load

3. Increase file transfer buffer from 512 to 16384 bytes

The 512-byte buffer size results in excessive system calls when serving
files. For a 1MB file, this requires 2048 read/write operations.

Using 16KB reduces system call overhead by 32x and better matches typical
OS page sizes and network buffer defaults. This should improve throughput
for static file serving while maintaining reasonable stack usage.

jart commented on llamafile/server/client.cpp:
```diff
@@ -520,14 +520,23 @@ Client::send_response_finish()
+        if (sent == -1) {
+            if (errno == EAGAIN || errno == EWOULDBLOCK) {
+                // no data can be written right now; retry
+                continue;
```
The right thing to do here would be to call poll() I think to wait until it's writable.

jart commented on llamafile/server/worker.cpp:
```diff
@@ -56,13 +56,6 @@ Worker::begin()
         tokens = tokenbucket_acquire(client_.client_ip_);
     server_->lock();
     dll_remove(&server_->idle_workers, &elem_);
-    if (dll_is_empty(server_->idle_workers)) {
```
Have you considered increasing the maximum number of threads? One of the design goals with the server is to have it be able to manage its resources. I'd ideally like to see us doing it in a better way than simply giving up on this.

jart replied:
Maybe a good compromise here would be this: we can add a flag to disable this. That will get you unblocked fastest.

vlasky replied:
I appreciate the design goal of active resource management, but I believe this specific approach is fundamentally flawed. The problem isn't thread count. It's this:

- 100 threads, all busy -> new connection kills active one
- 1000 threads, all busy -> new connection kills active one

No matter the thread count, under load we prioritise *new* connections over *active* ones. A connection 90% through a 10 MB transfer gets killed for one that hasn't done anything yet.

This seems like a Denial-of-Service vector - an attacker can kill N legitimate in-progress transfers by opening N new connections. The attack cost is trivial (TCP SYN) while the disruption is significant.

The TCP backlog IS resource management - I wouldn't say removing this code is "giving up" - it's using the OS's connection queue. The kernel holds new connections until a worker is free, and rejects them with TCP RST if the backlog fills. This is battle-tested, first-come-first-served resource management.

The current design suggest 'newest connection wins', which doesn't seem right. I'd prefer not to add a flag to preserve behavior I believe is incorrect. Happy to discuss further if you see a flaw in this reasoning.

vlasky commented:
Commit 521f7b744dd3c6dde842f58dbf91ec766c2c0d49 addresses remaining issues with socket write handling that weren't fully solved by the previous partial write fix (item 2).

Problems addressed:

1. Busy-loop on EAGAIN/EWOULDBLOCK: When write() returned EAGAIN/EWOULDBLOCK (socket buffer full), the code would immediately retry in a tight loop, wasting CPU cycles. Now we use poll() on the socket to efficiently wait for it to become writable, as recommended. We have a new setting --http-write-timeout flag (default: 60 seconds), which is the timeout for socket write operations.

2. Missing EINTR handling: Signal-interrupted system calls weren't being retried, which could cause spurious connection failures.
  
3. Incomplete partial write handling in safe_writev(): The vectored I/O function only called writev() once and didn't handle partial writes across multiple iovecs. Now we properly handle this by tracking progress across the iovec array and retrying until all data is sent or an error occurs.
